compiler/rustc_mir_build/src/thir/constant.rs

```diff
@@ -47,7 +47,7 @@ crate fn lit_to_const<'tcx>(
             trunc(if neg { (*n as i128).overflowing_neg().0 as u128 } else { *n })?
         }
         (ast::LitKind::Float(n, _), ty::Float(fty)) => {
-            parse_float(*n, *fty, neg).map_err(|_| LitToConstError::UnparseableFloat)?
+            parse_float(*n, *fty, neg).map_err(|_| LitToConstError::Reported)?
```

Comment:
With this PR, can the lexer still generate a float token that would fail in `parse_float`? Otherwise I'd expect to see a panic instead of an `Err`: in the `parse_function` below, replace `num.parse().map_err(|_| ())?` with a panic and return `ConstValue<'tcx>` instead of `Result<ConstValue<'tcx>, ()>`.

Reply:
I'm checking now, give me a bit...

Reply:
Short answer, no, it cannot fail:

Ok I think I've found where the lexer tokenizes the float. In `compiler/rustc_lexer/src/lib.rs`, we have:

```rust
    fn eat_decimal_digits(&mut self) -> bool {
        let mut has_digits = false;
        loop {
            match self.first() {
                '_' => {
                    self.bump();
                }
                '0'..='9' => {
                    has_digits = true;
                    self.bump();
                }
                _ => break,
            }
        }
        has_digits
    }

    fn eat_hexadecimal_digits(&mut self) -> bool {
        let mut has_digits = false;
        loop {
            match self.first() {
                '_' => {
                    self.bump();
                }
                '0'..='9' | 'a'..='f' | 'A'..='F' => {
                    has_digits = true;
                    self.bump();
                }
                _ => break,
            }
        }
        has_digits
    }

    /// Eats the float exponent. Returns true if at least one digit was met,
    /// and returns false otherwise.
    fn eat_float_exponent(&mut self) -> bool {
        debug_assert!(self.prev() == 'e' || self.prev() == 'E');
        if self.first() == '-' || self.first() == '+' {
            self.bump();
        }
        self.eat_decimal_digits()
    }

    fn number(&mut self, first_digit: char) -> LiteralKind {
        debug_assert!('0' <= self.prev() && self.prev() <= '9');
        let mut base = Base::Decimal;
        if first_digit == '0' {
            // Attempt to parse encoding base.
            let has_digits = match self.first() {
                'b' => {
                    base = Base::Binary;
                    self.bump();
                    self.eat_decimal_digits()
                }
                'o' => {
                    base = Base::Octal;
                    self.bump();
                    self.eat_decimal_digits()
                }
                'x' => {
                    base = Base::Hexadecimal;
                    self.bump();
                    self.eat_hexadecimal_digits()
                }
                // Not a base prefix.
                '0'..='9' | '_' | '.' | 'e' | 'E' => {
                    self.eat_decimal_digits();
                    true
                }
                // Just a 0.
                _ => return Int { base, empty_int: false },
            };
            // Base prefix was provided, but there were no digits
            // after it, e.g. "0x".
            if !has_digits {
                return Int { base, empty_int: true };
            }
        } else {
            // No base prefix, parse number in the usual way.
            self.eat_decimal_digits();
        };

        match self.first() {
            // Don't be greedy if this is actually an
            // integer literal followed by field/method access or a range pattern
            // (`0..2` and `12.foo()`)
            '.' if self.second() != '.' && !is_id_start(self.second()) => {
                // might have stuff after the ., and if it does, it needs to start
                // with a number
                self.bump();
                let mut empty_exponent = false;
                if self.first().is_digit(10) {
                    self.eat_decimal_digits();
                    match self.first() {
                        'e' | 'E' => {
                            self.bump();
                            empty_exponent = !self.eat_float_exponent();
                        }
                        _ => (),
                    }
                }
                Float { base, empty_exponent }
            }
            'e' | 'E' => {
                self.bump();
                let empty_exponent = !self.eat_float_exponent();
                Float { base, empty_exponent }
            }
            _ => Int { base, empty_int: false },
        }
    }
```

Likewise, in `compiler/rustc_parse/src/lexer/mod.rs` we have:

```rust
            rustc_lexer::LiteralKind::Float { base, empty_exponent } => {
                if empty_exponent {
                    self.err_span_(start, self.pos, "expected at least one digit in exponent");
                }

                match base {
                    Base::Hexadecimal => self.err_span_(
                        start,
                        suffix_start,
                        "hexadecimal float literal is not supported",
                    ),
                    Base::Octal => {
                        self.err_span_(start, suffix_start, "octal float literal is not supported")
                    }
                    Base::Binary => {
                        self.err_span_(start, suffix_start, "binary float literal is not supported")
                    }
                    _ => (),
                }

                let id = self.symbol_from_to(start, suffix_start);
                return (token::Float, id);
            }
```

Finally, in `compiler/rustc_parse/src/parser/expr.rs`, we handle cases where the leading character is a `.`:

```rust
    /// Matches `lit = true | false | token_lit`.
    /// Returns `None` if the next token is not a literal.
    pub(super) fn parse_opt_lit(&mut self) -> Option<Lit> {
        let mut recovered = None;
        if self.token == token::Dot {
            // Attempt to recover `.4` as `0.4`. We don't currently have any syntax where
            // dot would follow an optional literal, so we do this unconditionally.
            recovered = self.look_ahead(1, |next_token| {
                if let token::Literal(token::Lit { kind: token::Integer, symbol, suffix }) =
                    next_token.kind
                {
                    if self.token.span.hi() == next_token.span.lo() {
                        let s = String::from("0.") + &symbol.as_str();
                        let kind = TokenKind::lit(token::Float, Symbol::intern(&s), suffix);
                        return Some(Token::new(kind, self.token.span.to(next_token.span)));
                    }
                }
                None
            });
            if let Some(token) = &recovered {
                self.bump();
                self.error_float_lits_must_have_int_part(&token);
            }
        }
```

So what does this mean? Well, the logic is pretty clear:

In `/compiler/rustc_parse/src/lexer/mod.rs`, we can see that `LiteralKind::Float` cannot have an empty exponent, so there must be digits afterwards if there is a sign, and it must have a decimal base. So, this already eliminates a lot of unusual floats, such as `0b0101.101`.

In `compiler/rustc_lexer/src/lib.rs`, we have the following logic, bottom-up:

- **eat_decimal_digits:** Only matches `[0-9_]`.
- **eat_float_exponent:** Has to match `[eE][-+]?[0-9_]*`.
- **number:** A bit more complex, but quite simple.
  - If the leading character are digits, consume them.
  - If the next character is a `.`, consume optional digits afterwards.
    - If the next character is `[eE]`, consume an exponent.
  - If the next character is `[eE]`, consume an exponent.
- **parse_opt_lit:** If the next token is a float, and there's no duplicate `.`, then create a float token.

Reply:
Fixed as of [870ed2163282b2e5fa99585e6c28c916e4861646](https://github.com/Alexhuszagh/rust/blob/870ed2163282b2e5fa99585e6c28c916e4861646/compiler/rustc_mir_build/src/thir/constant.rs).

```rust
crate fn lit_to_const<'tcx>(
    tcx: TyCtxt<'tcx>,
    lit_input: LitToConstInput<'tcx>,
) -> Result<&'tcx ty::Const<'tcx>, LitToConstError> {
    ...
    let lit = match (lit, &ty.kind()) {
        (ast::LitKind::Float(n, _), ty::Float(fty)) => parse_float(*n, *fty, neg),
        ...
    }
    ...
}

fn parse_float<'tcx>(num: Symbol, fty: ty::FloatTy, neg: bool) -> ConstValue<'tcx> {
     ...
     let scalar = match fty {
        ty::FloatTy::F32 => {
            let rust_f = num.parse::<f32>().unwrap_or_else(|e| {
                panic!("apfloat::ieee::f32 failed to parse `{}`: {:?}", num, e)
            });
            ...
       }
       ty::FloatTy::F64 => {
            let rust_f = num.parse::<f64>().unwrap_or_else(|e| {
                panic!("apfloat::ieee::f64 failed to parse `{}`: {:?}", num, e)
            });
            ...
       }
    };

    ConstValue::Scalar(scalar)
}
```